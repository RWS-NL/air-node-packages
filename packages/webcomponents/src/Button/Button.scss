@import '../styles/vars';

.button {
  @extend %fontNormal;
  background-color: $skyBlue100;
  border-radius: 0;
  float: right;
  height: 3.2rem;
  line-height: initial;
  margin-right: 1%;
  text-transform: none;

  &Label {
    @extend %fontNormal;
    color: $white;
    font-size: 1rem;
    text-transform: none;
  }

  &Secondary {
    background-color: $yellow100;
    color: $textBlack;

    > span {
      color: $textBlack;
    }
  }

  &Outlined {
    background-color: $white;
    color: $skyBlue100;

    > span {
      color: $skyBlue100;
    }

    &Secondary {
      background-color: $white;
      color: $textBlack;

      > span {
        color: $textBlack;
      }
    }
  }

  &Disabled {
    background-color: $grey35 !important;
    border-color: $grey35 !important;
    box-shadow: none !important;
  }
}
