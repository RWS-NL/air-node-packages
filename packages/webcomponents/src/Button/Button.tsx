import MUIButton, { ButtonProps as MUIButtonProps } from '@material-ui/core/Button';
import css from './Button.scss';
import classnames from 'classnames';
import React, { FC, ReactNode, memo } from 'react';
import { makeStyles, Theme, createStyles } from '@material-ui/core';

export interface ButtonProps {
  /** The variant of the button */
  variant: 'text' | 'outlined' | 'contained' | undefined;
  /** The color type of the button */
  color: 'primary' | 'secondary' | 'default' | 'inherit' | undefined;
  /** The label to be displayed inside the button */
  label: ReactNode;
  /** Whether this button should be disabled */
  disabled?: boolean;
  /** Data-qa tag to apply to the search bar and input element */
  'data-qa'?: string;
  /** Custom CSS classes to pass to the button */
  customclasses?: string | string[];
  /** Custom CSS classes to pass to the button label */
  customlabelclasses?: string | string[];
  /** The action that should be triggered when clicking the button */
  onClick(): unknown;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({ buttonShadow: { boxShadow: theme.shadows[1], border: 'transparant' } })
);

/**
 * Constructs a button using pre-defined Rijkswaterstaat styling
 * @param props Props to pass to the button
 * @example
 * <Button label='button' onClick={() => undefined} variant='contained' color='primary' />
 */
export const Button: FC<MUIButtonProps<'button', ButtonProps>> = props => {
  const classes = useStyles();

  return (
    <MUIButton
      {...props}
      data-qa={props['data-qa']}
      onClick={props.onClick}
      variant={props.variant}
      color={props.color}
      disabled={props.disabled}
      className={classnames(props.customclasses)}
      classes={{
        root: css.button,
        label: classnames(css.buttonLabel, props.customlabelclasses),
        disabled: css.buttonDisabled,
        containedPrimary: classnames(css.buttonPrimary, classes.buttonShadow),
        containedSecondary: classnames(css.buttonSecondary, classes.buttonShadow),
        outlined: classnames(css.buttonOutlined, classes.buttonShadow),
        outlinedPrimary: classnames(css.buttonOutlined, classes.buttonShadow),
        outlinedSecondary: classnames(css.buttonOutlinedSecondary, classes.buttonShadow)
      }}
      href={undefined}
    >
      {props.label}
    </MUIButton>
  );
};

export default memo(Button);
