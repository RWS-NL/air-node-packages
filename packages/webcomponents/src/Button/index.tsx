import { createStyles, makeStyles, Theme } from '@material-ui/core';
import MUIButton, { ButtonProps as MUIButtonProps } from '@material-ui/core/Button';
import classnames from 'classnames';
import React, { memo, ReactNode } from 'react';
import css from './Button.scss';

export interface ButtonProps {
  /** The variant of the button */
  variant: 'text' | 'outlined' | 'contained' | undefined;
  /** The color type of the button */
  color: 'primary' | 'secondary' | 'default' | 'inherit' | undefined;
  /** The label to be displayed inside the button */
  label: ReactNode;
  /** Whether this button should be disabled */
  disabled?: boolean;
  /** Data-qa tag to apply to the search bar and input element */
  'data-qa'?: string;
  /** Custom CSS classes to pass to the button */
  customclasses?: string | string[];
  /** Custom CSS classes to pass to the button label */
  customlabelclasses?: string | string[];
  /** The action that should be triggered when clicking the button */
  onClick(): unknown;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({ buttonShadow: { boxShadow: theme.shadows[1], border: 'transparant' } })
);

/**
 * Constructs a button using pre-defined Rijkswaterstaat styling
 * @param props Props to pass to the button
 * @example
 * ```jsx
 * <Button label='button' onClick={() => undefined} variant='contained' color='primary' />
 * ```
 */
export const Button = memo((props: MUIButtonProps<'button', ButtonProps>) => {
  const classes = useStyles();

  return (
    <MUIButton
      {...props}
      data-qa={props['data-qa']}
      onClick={props.onClick}
      variant={props.variant}
      color={props.color}
      disabled={props.disabled}
      className={classnames(props.customclasses)}
      classes={{
        root: css.button,
        label: classnames(css.buttonLabel, props.customlabelclasses),
        disabled: css.buttonDisabled,
        containedPrimary: classnames(css.buttonPrimary, classes.buttonShadow),
        containedSecondary: classnames(css.buttonSecondary, classes.buttonShadow),
        outlined: classnames(css.buttonOutlined, classes.buttonShadow),
        outlinedPrimary: classnames(css.buttonOutlined, classes.buttonShadow),
        outlinedSecondary: classnames(css.buttonOutlinedSecondary, classes.buttonShadow)
      }}
      href={undefined}
    >
      {props.label}
    </MUIButton>
  );
});

export default Button;
