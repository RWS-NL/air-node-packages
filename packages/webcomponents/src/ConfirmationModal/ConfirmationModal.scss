@import '../styles/vars';

%modalButton {
  width: 11.5rem;
}

.modal {
  > div > div {
    border-radius: 0;
    max-width: unset;
    min-width: 45vw;
  }

  &Actions {
    margin-bottom: 3vh;
    margin-right: 5vw;
    padding-right: 1.5rem;
  }

  &Body {
    margin: 0 5vw;
    overflow: hidden;

    &Small {
      max-height: 10rem;
    }
  }

  &Button {
    &Cancel {
      @extend %modalButton;
      background-color: $white;
      border-width: 2px;

      > span {
        color: $skyBlue100;
      }
    }

    &Ok {
      @extend %modalButton;
      margin-left: 1%;
    }
  }
}

.title {
  margin-left: 5vw;
  margin-right: 5vw;
  padding-top: 3vw;

  > * {
    align-content: center;
    color: $logoBlue;
    display: flex;
    font-size: 1.1rem;
    font-weight: bolder;
  }

  &Icon {
    box-sizing: content-box;
    padding-bottom: 0.6rem;
    padding-right: 0.8rem;

    &CheckCircle {
      color: $green100;
    }

    &Offset {
      margin-left: calc(5vw - 2.5rem);
    }
  }
}
