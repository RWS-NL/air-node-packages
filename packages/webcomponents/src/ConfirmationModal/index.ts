export { default as ConfirmationModal, ConfirmationModalProps, ConfirmationModalQAs } from './ConfirmationModal';