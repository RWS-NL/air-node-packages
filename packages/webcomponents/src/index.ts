export * from './ActionBar';
export * from './Button';
export * from './ConfirmationModal';
export * from './constants';
export * from './Dropdownbutton';
export * from './LoadingSkeleton';
export * from './Logo';
export * from './Modal';
export * from './ModalContent';
export * from './PaperButton';
export * from './SearchBar';
export * from './Table';
export * from './TableBodyCell';
export * from './TableHeaderCell';
export * from './TablePagination';
export * from './TablePaginationActions';
export * from './TableToolbar';
export * from './Tooltip';
export * from './ZeroWidthSpace';