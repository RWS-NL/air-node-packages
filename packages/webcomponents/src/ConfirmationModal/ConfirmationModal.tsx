import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import CheckCircle from '@material-ui/icons/CheckCircle';
import WarningIcon from '@material-ui/icons/Warning';
import { objectHasProperty } from '@rws-air/utils';
import classnames from 'classnames';
import { ModalProps, ModalQAs } from '../Modal/Modal';
import React, { FC, Fragment, KeyboardEvent, memo, useMemo } from 'react';
import Button from '../Button/Button';
import { dataQa } from '../constants';
import css from './ConfirmationModal.scss';

export interface ConfirmationModalQAs extends ModalQAs {
  /** Data-qa applied to the modal actions */
  actions?: dataQa;
  /** Data-qa applied to the action confirm button */
  actionConfirm?: dataQa;
  /** Data-qa applied to the action cancel button */
  actionCancel?: dataQa;
}

export interface ConfirmationModalProps extends ModalProps {
  /** Object of data-qa tags to pass to the modal */
  modalqas?: ConfirmationModalQAs;
  /** Text to show in the confirm button */
  okButtonText: string;
  /** Text to show in the cancel button */
  cancelButtonText: string;
  /** Action performed when the cancel button is clicked */
  closeAction: () => unknown;
  /** Action performed when the confirm button is clicked */
  confirmAction: () => unknown;
  /** Whether the icon for this modal should be a Warning triangle or a Check Circle */
  modalType: 'warning' | 'check';
  /** Whether clicking on the backdrop should close the modal (triggers closeAction)  */
  disableBackdropClick?: boolean;
  /** Whether pressing the escape key should close the modal (triggers closeAction) */
  disableEscapeKeyDown?: boolean;
}

/**
 * Creates a modal using pre-defined Rijkswaterstaat styling
 */
export const ConfirmationModal: FC<ConfirmationModalProps> = props => {
  const handleConfirmKeyPress = (event: KeyboardEvent) => {
    return props.confirmAction && event.key.toLowerCase() === 'enter'
      ? props.confirmAction()
      : undefined;
  };

  const handleConfirm = () => props.confirmAction();
  const handleClose = () => props.closeAction();

  const renderModalActions = useMemo(
    () => {
      return (
        <DialogActions classes={{ root: css.modalActions }} data-qa={props.modalqas?.actions || 'modal-actions'}>
          <Button
            data-qa={props.modalqas?.actionCancel || 'modal-cancel-button'}
            onClick={handleClose} variant='outlined' color='primary'
            label={props.cancelButtonText ? props.cancelButtonText : 'cancel'}
            customclasses={css.modalButtonCancel}
          />
          <Button
            data-qa={props.modalqas?.actionConfirm || 'modal-confirm-button'}
            onClick={handleConfirm} variant='contained' color='primary'
            label={props.okButtonText ? props.okButtonText : 'ok'}
            customclasses={css.modalButtonOk}
          />
        </DialogActions>
      );
    },
    [
      handleClose, handleConfirm,
      props.cancelButtonText, props.okButtonText
    ]
  );

  const renderModalTitleIcon = useMemo(() => {
    if (props.modalType === 'warning') return <WarningIcon color='error' className={classnames(css.titleIcon)} />;

    return <CheckCircle color='inherit' className={classnames(css.titleIcon, css.titleIconCheckCircle)} />;
  }, [ props.modalType ]);

  return (
    <Dialog
      disableBackdropClick={objectHasProperty(props, 'disableBackdropClick') ? props.disableBackdropClick : true}
      disableEscapeKeyDown={objectHasProperty(props, 'disableEscapeKeyDown') ? props.disableEscapeKeyDown : true}
      open={Boolean(props.open)}
      classes={{ root: css.modal }}
      onKeyPress={handleConfirmKeyPress}
      data-qa={props.modalqas?.modal || 'modal'}
    >
      <DialogTitle
        data-qa={props.modalqas?.title || 'modal-title'}
        className={classnames(
          css.title,
          css.titleIconOffset,
          css.titleIconOffsetHeader
        )}
      >
        <Fragment>
          {renderModalTitleIcon}
          {props.topic}
        </Fragment>
      </DialogTitle>
      <DialogContent
        classes={{ root: css.modalBody }}
        className={css.modalContentSmall}
        data-qa={props.modalqas?.content || 'modal-content'}
      >
        {props.children}
      </DialogContent>
      {renderModalActions}
    </Dialog>
  );
};

export default memo(ConfirmationModal);