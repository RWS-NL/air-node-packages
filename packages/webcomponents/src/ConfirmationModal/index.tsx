import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import CheckCircle from '@material-ui/icons/CheckCircle';
import WarningIcon from '@material-ui/icons/Warning';
import classnames from 'classnames';
import React, { Fragment, KeyboardEvent, memo, useMemo } from 'react';
import Button from '../Button';
import { ModalProps, ModalQAs } from '../Modal';
import css from './ConfirmationModal.scss';

export interface ConfirmationModalQAs extends ModalQAs {
  /** Data-qa applied to the modal actions */
  actions?: string;
  /** Data-qa applied to the action confirm button */
  actionConfirm?: string;
  /** Data-qa applied to the action cancel button */
  actionCancel?: string;
}

export interface ConfirmationModalProps extends ModalProps {
  /** Object of data-qa tags to pass to the modal */
  modalqas?: ConfirmationModalQAs;
  /** Text to show in the confirm button */
  okButtonText: string;
  /** Text to show in the cancel button */
  cancelButtonText: string;
  /** Whether the icon for this modal should be a Warning triangle or a Check Circle */
  modalType: 'warning' | 'check';

  /** Action performed when the cancel button is clicked */
  closeAction(): void;
  /** Action performed when the confirm button is clicked */
  confirmAction(): void;
}

/**
 * Constructs a confirmation modal using pre-defined Rijkswaterstaat styling
 * @param props Props to pass to the confirmation modal
 * @example
 * ```jsx
 * <ConfirmationModal
 *   modalType='warning'
 *   open={true}
 *   topic='SAMPLE TOPIC'
 *   cancelButtonText='CANCEL'
 *   okButtonText='OK'
 *   confirmAction={console.log}
 *   closeAction={console.log}
 *   modalqas={{
 *      modal: 'sample',
 *      content: 'sample-content',
 *      title: 'sample-title',
 *      actions: 'sample-actions',
 *      actionCancel: 'sample-action-cancel',
 *      actionConfirm: 'sample-action-confirm',
 *   }}
 *   dialogContent={ <ModalContent><div>SAMPLE</div></ModalContent> }
 * />
 * ```
 */
export const ConfirmationModal = memo((props: ConfirmationModalProps) => {
  const handleConfirmKeyPress = (event: KeyboardEvent) => {
    return props.confirmAction && event.key.toLowerCase() === 'enter' ? props.confirmAction() : undefined;
  };

  const renderModalActions = useMemo(() => {
    return (
      <DialogActions classes={{ root: css.modalActions }} data-qa={props.modalqas?.actions || 'modal-actions'}>
        <Button
          data-qa={props.modalqas?.actionCancel || 'modal-cancel-button'}
          onClick={props.closeAction}
          variant='outlined'
          color='primary'
          label={props.cancelButtonText ? props.cancelButtonText : 'cancel'}
          customclasses={css.modalButtonCancel}
        />
        <Button
          data-qa={props.modalqas?.actionConfirm || 'modal-confirm-button'}
          onClick={props.confirmAction}
          variant='contained'
          color='primary'
          label={props.okButtonText ? props.okButtonText : 'ok'}
          customclasses={css.modalButtonOk}
        />
      </DialogActions>
    );
  }, [props.modalqas, props.closeAction, props.cancelButtonText, props.confirmAction, props.okButtonText]);

  const renderModalTitleIcon = useMemo(() => {
    if (props.modalType === 'warning') return <WarningIcon color='error' className={classnames(css.titleIcon)} />;

    return <CheckCircle color='inherit' className={classnames(css.titleIcon, css.titleIconCheckCircle)} />;
  }, [props.modalType]);

  return (
    <Dialog
      disableBackdropClick
      disableEscapeKeyDown
      open={Boolean(props.open)}
      classes={{ root: css.modal }}
      onKeyPress={handleConfirmKeyPress}
      data-qa={props.modalqas?.modal || 'modal'}
    >
      <DialogTitle
        data-qa={props.modalqas?.title || 'modal-title'}
        className={classnames(css.title, css.titleIconOffset)}
      >
        <Fragment>
          {renderModalTitleIcon}
          {props.topic}
        </Fragment>
      </DialogTitle>
      <DialogContent
        classes={{ root: css.modalBody }}
        className={css.modalContentSmall}
        data-qa={props.modalqas?.content || 'modal-content'}
      >
        {props.dialogContent}
      </DialogContent>
      {renderModalActions}
    </Dialog>
  );
});

export default ConfirmationModal;
