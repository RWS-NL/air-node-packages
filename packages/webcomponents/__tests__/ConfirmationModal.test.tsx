import CheckCircle from '@material-ui/icons/CheckCircle';
import WarningIcon from '@material-ui/icons/Warning';
import { shallow, ShallowWrapper } from 'enzyme';
import React, { Fragment } from 'react';
import Button from '../src/Button';
import { ConfirmationModal, ConfirmationModalProps } from '../src/ConfirmationModal';

let confirmationModal: ShallowWrapper<ConfirmationModalProps, any>;
const closeAction = jest.fn();
const confirmAction = jest.fn();

const setup = (
  isOpen = false,
  props?: Partial<ConfirmationModalProps>,
  DialogContent: () => JSX.Element = () => <Fragment />
) =>
  (confirmationModal = shallow<ConfirmationModalProps>(
    <ConfirmationModal
      topic='Awesome Topic'
      okButtonText='OK'
      cancelButtonText='CANCEL'
      modalType='warning'
      closeAction={closeAction}
      confirmAction={confirmAction}
      open={isOpen}
      dialogContent={<DialogContent />}
      modalqas={{
        content: 'modal-content',
        modal: 'modal',
        title: 'modal-title',
        actionCancel: 'modal-cancel-button',
        actionConfirm: 'modal-confirm-button',
        actions: 'modal-actions'
      }}
      {...props}
    />
  ));

describe('Render Testing', () => {
  test('should include mandatory props', () => {
    setup();
    expect(confirmationModal.find(Button)).toHaveLength(2);
    expect(confirmationModal.prop('disableBackdropClick')).toBe(true);
    expect(confirmationModal.prop('disableEscapeKeyDown')).toBe(true);
  });

  test("should pass all data-qa's", () => {
    setup(true);

    expect(confirmationModal.find('[data-qa="modal"]').exists()).toBe(true);
    expect(confirmationModal.find('[data-qa="modal-title"]').exists()).toBe(true);
    expect(confirmationModal.find('[data-qa="modal-cancel-button"]').exists()).toBe(true);
    expect(confirmationModal.find('[data-qa="modal-confirm-button"]').exists()).toBe(true);
    expect(confirmationModal.find('[data-qa="modal-actions"]').exists()).toBe(true);
  });

  test('should show CheckCircle when modalType == check', () => {
    setup(true, { modalType: 'check' });

    expect(confirmationModal.exists()).toBe(true);
    expect(confirmationModal.find(CheckCircle).exists()).toBe(true);
  });

  test('should show Warning Triangle when modalType == check', () => {
    setup(true, { modalType: 'warning' });

    expect(confirmationModal.find(WarningIcon).exists()).toBe(true);
  });
});

describe('Action Testing', () => {
  beforeAll(() => setup(true));
  test('Ok button should work', () => {
    const okButton = confirmationModal.find('[data-qa="modal-confirm-button"]');
    okButton.simulate('click');
    expect(confirmAction).toHaveBeenCalledWith();
    expect(confirmAction).toHaveBeenCalledTimes(1);
  });

  test('Cancel button should work', () => {
    const cancelButton = confirmationModal.find('[data-qa="modal-cancel-button"]');
    cancelButton.simulate('click');
    expect(closeAction).toHaveBeenCalledWith();
    expect(closeAction).toHaveBeenCalledTimes(1);
  });
});

describe('Content Testing', () => {
  const Content = () => <div>Content</div>;
  beforeAll(() => setup(true, {}, Content));

  test('should contain content', () => {
    expect(
      confirmationModal
        .find('[data-qa="modal-content"]')
        .children()
        .exists()
    ).toBe(true);
  });
});

describe('Snapshot Testing', () => {
  test('Required Props', () => {
    setup();
    expect(confirmationModal).toMatchSnapshot();
  });
});
