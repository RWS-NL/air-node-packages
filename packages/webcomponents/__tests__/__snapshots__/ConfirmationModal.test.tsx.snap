// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing Required Props 1`] = `
<WithStyles(ForwardRef(Dialog))
  classes={
    Object {
      "root": "modal",
    }
  }
  data-qa="modal"
  disableBackdropClick={true}
  disableEscapeKeyDown={true}
  onKeyPress={[Function]}
  open={false}
>
  <WithStyles(ForwardRef(DialogTitle))
    className="title titleIconOffset titleIconOffsetHeader"
    data-qa="modal-title"
  >
    <WarningIcon
      className="titleIcon"
      color="error"
    />
    Awesome Topic
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))
    className="modalContentSmall"
    classes={
      Object {
        "root": "modalBody",
      }
    }
    data-qa="modal-content"
  >
    <Component />
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(DialogActions))
    classes={
      Object {
        "root": "modalActions",
      }
    }
    data-qa="modal-actions"
  >
    <Memo(Button)
      color="primary"
      customclasses="modalButtonCancel"
      data-qa="modal-cancel-button"
      label="CANCEL"
      onClick={[Function]}
      variant="outlined"
    />
    <Memo(Button)
      color="primary"
      customclasses="modalButtonOk"
      data-qa="modal-confirm-button"
      label="OK"
      onClick={[Function]}
      variant="contained"
    />
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;
